src/library/scala/collection/immutable/TreeSeqMap.scala

```diff
@@ -0,0 +1,638 @@
+        while (rearIter.hasNext) {
+          mng = mng - rearIter.next()
+        }
+        new TreeSeqMap(ong, mng, ordinal, orderedBy)
+      } else {
+        // Populate with builder otherwise
+        val bdr = newBuilder[K, V](orderedBy)
+        val iter = ordering.iterator
+        var i = 0
+        while (i < f) {
+          iter.next()
+          i += 1
+        }
+        while (i < u) {
+          val k = iter.next()
+          bdr.addOne(k, mapping(k)._2)
+          i += 1
+        }
+        bdr.result()
+      }
+    }
+  }
+
+  override def map[K2, V2](f: ((K, V)) => (K2, V2)): TreeSeqMap[K2, V2] = {
+    val bdr = newBuilder[K2, V2](orderedBy)
+    val iter = ordering.iterator
+    while (iter.hasNext) {
+      val k = iter.next()
+      val (_, v) = mapping(k)
+      val (k2, v2) = f((k, v))
+      bdr.addOne(k2, v2)
+    }
+    bdr.result()
+  }
+
+  override def flatMap[K2, V2](f: ((K, V)) => IterableOnce[(K2, V2)]): TreeSeqMap[K2, V2] = {
+    val bdr = newBuilder[K2, V2](orderedBy)
+    val iter = ordering.iterator
+    while (iter.hasNext) {
+      val k = iter.next()
+      val (_, v) = mapping(k)
+      val jter = f((k, v)).iterator
+      while (jter.hasNext) {
+        val (k2, v2) = jter.next()
+        bdr.addOne(k2, v2)
+      }
+    }
+    bdr.result()
+  }
+
+  override def collect[K2, V2](pf: PartialFunction[(K, V), (K2, V2)]): TreeSeqMap[K2, V2] = {
+    val bdr = newBuilder[K2, V2](orderedBy)
+    val iter = ordering.iterator
+    while (iter.hasNext) {
+      val k = iter.next()
+      val (_, v) = mapping(k)
+      if (pf.isDefinedAt((k, v))) {
+        val (k2, v2) = pf((k, v))
+        bdr.addOne(k2, v2)
+      }
+    }
+    bdr.result()
+  }
+
+  override def concat[V2 >: V](suffix: IterableOnce[(K, V2)]): TreeSeqMap[K, V2] = {
+    var ong: Ordering[K] = ordering
+    var mng: Mapping[K, V2] = mapping
+    var ord = increment(ordinal)
+    val iter = suffix.iterator
+    while (iter.hasNext) {
+      val (k, v2) = iter.next()
+      mng.get(k) match {
+        case Some((o, v)) =>
+          if (orderedBy == OrderBy.Insertion && v != v2) mng = mng.updated(k, (o, v2))
+          else if (orderedBy == OrderBy.Modification) {
+            mng = mng.updated(k, (ord, v2))
+            ong = ong.exclude(o).append(ord, k)
+            ord = increment(ord)
+          }
+        case None =>
+          mng = mng.updated(k, (ord, v2))
+          ong = ong.append(ord, k)
+          ord = increment(ord)
+      }
+    }
+    new TreeSeqMap[K, V2](ong, mng, ord, orderedBy)
+  }
+
+  @`inline` private[this] def value(p: (_, V)) = p._2
+  @`inline` private[this] def binding(k: K) = mapping(k).copy(_1 = k)
+}
+object TreeSeqMap extends MapFactory[TreeSeqMap] {
+  sealed trait OrderBy
+  final object OrderBy {
+    final case object Insertion extends OrderBy
+    final case object Modification extends OrderBy
+  }
+
+  val Empty = new TreeSeqMap[Nothing, Nothing](Ordering.empty, HashMap.empty, 0, OrderBy.Insertion)
```

Comment:
Would it make sense to start at Int.MinValue?

Reply:
(Sorry for the nitpick)

Reply:
Positive indexing is easier.  I'd leave it alone.

Reply:
@joshlemer @SethTisue I'll repeat the answer I gave to @NthPortal (https://github.com/scala/scala/pull/7146#discussion_r226823340) when he asked the same question: 

> The reason that the ordinal counting starts with `0: Int`, jumps from `Int.MaxValue` to `Int.MinValue` and repopulates a fresh instance once it reaches `-1: Int` is that the `IntMap` used for the ordering uses a tree structure that is based on the bit sequence of the `Int` keys. This means that `0: Int` (00000000000000000000000000000000) will be the leftmost (first) possible leaf in the tree, `Int.MaxValue` (01111111111111111111111111111111) will be followed directly by `Int.MinValue` (10000000000000000000000000000000), and the rightmost (last) possible leaf is `-1: Int` (11111111111111111111111111111111).
```scala
scala> IntMap(1 -> "one", 10 -> "ten", Int.MinValue -> "min", -10 -> "-ten", 0 -> "zero", -1 -> "-one", Int.MaxValue -> "max")
res55: scala.collection.immutable.IntMap[String] = IntMap(0 -> zero, 1 -> one, 10 -> ten, 2147483647 -> max, -2147483648 -> min, -10 -> -ten, -1 -> -one)
```